neural_lam/datastore/base.py

```diff
@@ -0,0 +1,371 @@
+
+    # Ensemble vs deterministic data
+    If the datastore is used to represent ensemble data, then the `is_ensemble`
+    attribute should be set to True, and returned data from `get_dataarray` is
+    assumed to have an `ensemble_member` dimension.
+
+    """
+
+    is_ensemble: bool = False
+    is_forecast: bool = False
+
+    @property
+    @abc.abstractmethod
+    def root_path(self) -> Path:
+        """
+        The root path to the datastore. It is relative to this that any derived
+        files (for example the graph components) are stored.
+
+        Returns
+        -------
+        pathlib.Path
+            The root path to the datastore.
+
+        """
+        pass
+
+    @property
+    @abc.abstractmethod
+    def config(self) -> collections.abc.Mapping:
+        """The configuration of the datastore.
+
+        Returns
+        -------
+        collections.abc.Mapping
+            The configuration of the datastore, any dict like object can be
+            returned.
+
+        """
+        pass
+
+    @property
+    @abc.abstractmethod
+    def step_length(self) -> int:
+        """The step length of the dataset in hours.
+
+        Returns:
+            int: The step length in hours.
+
+        """
+        pass
+
+    @abc.abstractmethod
+    def get_vars_units(self, category: str) -> List[str]:
+        """Get the units of the variables in the given category.
+
+        Parameters
+        ----------
+        category : str
+            The category of the variables.
+
+        Returns
+        -------
+        List[str]
+            The units of the variables.
+
+        """
+        pass
+
+    @abc.abstractmethod
+    def get_vars_names(self, category: str) -> List[str]:
+        """Get the names of the variables in the given category.
+
+        Parameters
+        ----------
+        category : str
+            The category of the variables.
+
+        Returns
+        -------
+        List[str]
+            The names of the variables.
+
+        """
+        pass
+
+    @abc.abstractmethod
+    def get_num_data_vars(self, category: str) -> int:
+        """Get the number of data variables in the given category.
+
+        Parameters
+        ----------
+        category : str
+            The category of the variables.
+
+        Returns
+        -------
+        int
+            The number of data variables.
+
+        """
+        pass
+
+    @abc.abstractmethod
+    def get_normalization_dataarray(self, category: str) -> xr.Dataset:
+        """
+        Return the normalization dataarray for the given category. This should
+        contain a `{category}_mean` and `{category}_std` variable for each
+        variable in the category. For `category=="state"`, the dataarray should
+        also contain a `state_diff_mean` and `state_diff_std` variable for the
+        one- step differences of the state variables. The returned dataarray
+        should at least have dimensions of `({categ ory}_feature)`, but can
```

Comment:
```suggestion
        should at least have dimensions of `({category}_feature)`, but can
```

Reply:
fixed in https://github.com/leifdenby/neural-lam/commit/8e7b2e6284609fb4d79eece6b816d7b553b1d94e